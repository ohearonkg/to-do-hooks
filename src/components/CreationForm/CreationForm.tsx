import React, { ChangeEvent, FormEvent } from "react";

interface ICreationFormProps {
  /**
   * Function to be called
   * when the user submits
   * the form
   */
  onSubmitFunction: (toDoText: string) => void;
}

interface ICreationFormState {
  /**
   * Current value of the input
   */
  inputValue: string;
}

export default class CreationForm extends React.Component<
  ICreationFormProps,
  ICreationFormState
> {
  public state = {
    inputValue: ""
  };

  /**
   * Handling updating use the event provided
   * via the change handler of the input
   */
  public handleTextUpdate = (event: ChangeEvent<HTMLInputElement>) =>
    this.setState({ inputValue: event.target.value });

  /**
   * Submitting our form
   */
  public handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    const { onSubmitFunction } = this.props;
    event.preventDefault();
    onSubmitFunction(this.state.inputValue);
  };

  public render() {
    const { inputValue } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          To Do:
          <input
            type="text"
            value={inputValue}
            onChange={this.handleTextUpdate}
          />
        </label>
        <button type="submit">ADD</button>
      </form>
    );
  }
}
