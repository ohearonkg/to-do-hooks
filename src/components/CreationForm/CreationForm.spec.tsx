import "jest-dom/extend-expect";

import { cleanup, fireEvent, render } from "react-testing-library";

import React from "react";
import CreationForm from "./CreationForm";

afterEach(cleanup);

describe("Creation Form", () => {
  /**
   * Snapshot
   */
  it("Should match its snapshot", () => {
    const { asFragment } = render(<CreationForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  /**
   * Calling its onSubmitFunction
   */
  it("Should call its onSubmitFunction when the form is submitted", () => {
    const sampleFunction = jest.fn();
    const { getByText } = render(
      <CreationForm onSubmitFunction={sampleFunction} />
    );
    fireEvent.click(getByText("ADD"));
    expect(sampleFunction).toHaveBeenCalledTimes(1);
  });

  /**
   * Passing text to onSubmitFunction
   */
  it("Should pass the text from its input to its onSubmitFunction when called", () => {
    const sampleToDoText = "BUY MILK";
    const sampleFunction = jest.fn();

    const { getByText, getByLabelText } = render(
      <CreationForm onSubmitFunction={sampleFunction} />
    );

    getByLabelText(/To Do/i).nodeValue = sampleToDoText;

    fireEvent.click(getByText("ADD"));
    expect(sampleFunction).toHaveBeenCalledWith(sampleToDoText);
  });
});
