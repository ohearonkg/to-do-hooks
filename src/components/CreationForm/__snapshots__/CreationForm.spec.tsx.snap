// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creation Form Should match its snapshot 1`] = `
<DocumentFragment>
  <h1>
     Hello 
  </h1>
</DocumentFragment>
`;
